### PR TITLE
Handle IAsyncDisposable scoped objects

### PR DESCRIPTION
**Pull request by davidfowl**
- We make a scope today around hub invocations, with IAsyncDisposable now implemented in the DI container, we need to support IServiceScope being IAsyncDisposable and IDisposable

https://github.com/aspnet/Extensions/commit/a38846cfa13a1e32a5aae90e54dbb71bef20a7db

### CODE REVIEW

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 1]
```diff
@@ -53,12 +57,28 @@ public override async Task OnConnectedAsync(HubConnectionContext connection)
                     hubActivator.Release(hub);
                 }
             }
+            finally
```
Comment (DAllanCarr) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Seems like alot of rinse repeat with the dispose pattern, could scope to method and inline.

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 2]
```diff
@@ -342,17 +358,18 @@ async Task ExecuteInvocation()
...
         }
 
-        private void CleanupInvocation(HubConnectionContext connection, HubMethodInvocationMessage hubMessage, IHubActivator<THub> hubActivator,
+        private async Task CleanupInvocation(HubConnectionContext connection, HubMethodInvocationMessage hubMessage, IHubActivator<THub> hubActivator,
             THub hub, IServiceScope scope)
         {
             hubActivator?.Release(hub);
-            scope.Dispose();
 
+            await scope.DisposeAsync();
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: I think it's safe to return the task at the end of the function so no statemachine.

```
var scopeDisposeTask = scope.DisposeAsync();
// ...
return scopeDisposeTask;
```

File: src/SignalR/server/Core/src/Internal/AsyncDisposableExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+using System;
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/Internal/AsyncDisposableExtensions.cs]: copyright

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 2 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Why?

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: To get rid of the statemachine

Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: It doesn't matter here. Not even a little bit

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 1 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: If you mean make a helper method that takes a lambda then that would have worse performance than just doing it inline.

Comment (DAllanCarr) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Extension method is sweet

**Review by BrennanConroy — APPROVED** [review]

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 2 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: OK did it 👍

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 4]
```diff
@@ -53,12 +57,20 @@ public override async Task OnConnectedAsync(HubConnectionContext connection)
                     hubActivator.Release(hub);
                 }
             }
+            finally
+            {
+                await scope.DisposeAsync();
```
Comment (halter73) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Will this be converted back to `async using`?

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs [hunk 5]
```diff
@@ -365,6 +378,10 @@ async Task ExecuteInvocation()
                     catch (KeyNotFoundException) { }
                 }
             }
+
+            hubActivator?.Release(hub);
```
Comment (halter73) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Why did you move `hubActivator?.Release(hub);` to the end? Is it important?

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 6]
```diff
@@ -54,6 +54,37 @@ public async Task HubsAreDisposed()
...
+
+                    var result = (await client.InvokeAsync(nameof(HubWithAsyncDisposable.Test)).OrTimeout());
+                    Assert.NotNull(result);
+
+                    // kill the connection
+                    client.Dispose();
+
+                    await connectionHandlerTask;
+
+                    Assert.Equal(3, trackDispose.DisposeCount);
```
Comment (halter73) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: Is this three because the HubConnectionHandler happened to resolve and dispose HubWithAsyncDisposable 3 times? That's MockBehavior.Strict levels of finicky 😆

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 5 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: No, I just wanted the Release and Dispose to be together.

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs (hunk 6 above)
Comment (BrennanConroy) [reply src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: 1. OnConnectedAsync
1. HubWithAsyncDisposable.Test
1. OnDisconnectedAsync

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs (hunk 4 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs]: Not unless we make a wrapper struct. Current plan is to leave it as is.

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs (hunk 6 above)
Comment (halter73) [reply src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: I know *why* the hub was resolved 3 times, but if we consider using the same hub for OnConnectedAsync and OnDisconnectedAsync is a bug, we should have a test that specifically tests that instead of making other tests unnecessarily strict. Not a big deal though, we'll just have one more test to change if we change this behavior.
